### PR TITLE
Ocp deployment using acm UI

### PR DESCRIPTION
**Pull request by shylesh**

### CODE REVIEW

**Comment by shylesh** [comment]
Rebased and conflict resolved

**Comment by shylesh** [comment]
Rebased on latest master

File: ocs_ci/deployment/multicluster_deployment.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        factory = acm_ui.ACMOCPDeploymentFactory()
+        self.ui_driver = acm.login_to_acm()
+
+        if self.deploy_sync_mode == "async":
+            rdr_clusters = get_non_acm_cluster_config()
+            for cluster_conf in rdr_clusters:
+                deployer = factory.get_platform_instance(self.ui_driver, cluster_conf)
+                deployer.create_cluster_prereq()
+                deployer.create_cluster()
+                self.deployment_status_list.append(deployer)
```
Comment (petr-balogh) [inline ocs_ci/deployment/multicluster_deployment.py:59]: typo , it should be deployment_cluster_list

**Comment by shylesh** [comment]
Rebased on latest master

**Comment by shylesh** [comment]
rebased on latest master

File: ocs_ci/framework/conf/default_config.yaml [hunk 2]
```diff
@@ -235,7 +235,7 @@ EXTERNAL_MODE:
 UI_SELENIUM:
   browser_type: "chrome"
   chrome_type: "google-chrome"
-  headless: True
+  headless: False
```
Comment (OdedViner) [inline ocs_ci/framework/conf/default_config.yaml:238]: Why did you change the headless to False?
We must to run it with headless on jenkins slave..

File: ocs_ci/ocs/ui/base_ui.py [hunk 3]
```diff
@@ -106,6 +106,10 @@ def do_send_keys(self, locator, text, timeout=30):
             logger.error(e)
             raise TimeoutException
 
+    def do_click_and_send_keys(self, locator, text, timeout=30):
```
Comment (OdedViner) [inline ocs_ci/ocs/ui/base_ui.py:109]: Why do you need "click" and "send_keys" on same method?

File: ocs_ci/ocs/constants.py [hunk 4]
```diff
@@ -1662,3 +1666,22 @@
...
+PLATFORM_XPATH_MAP = {
+    "vsphere": "cc_provider_vmware_vsphere",
+    "AWS": None,
+    "baremetal": None,
+    "azure": None,
+}
+ACM_PLATOFRM_VSPHERE_CRED_PREFIX = "vsphereacmocp-"
+# example release image url : quay.io/openshift-release-dev/ocp-release:4.9.23-x86_64
+ACM_OCP_RELEASE_IMG_URL_PREFIX = "registry.ci.openshift.org/ocp/release"
+ACM_VSPHERE_NETWORK = "".join(["VM", Keys.SPACE, "Network"])
```
Comment (petr-balogh) [inline ocs_ci/ocs/constants.py:1679]: As you described this doesn't work, we can have it as single string

File: ocs_ci/deployment/multicluster_deployment.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        if self.multicluster_mode == constants.RDR_MODE:
+            self.do_rdr_acm_ocp_deploy()
+
+    def do_rdr_acm_ocp_deploy(self):
+        """
+        Specific to regional DR OCP cluster deployments
+
+        """
+        factory = acm_ui.ACMOCPDeploymentFactory()
+        self.ui_driver = acm.login_to_acm()
```
Comment (am-agrawa) [inline ocs_ci/deployment/multicluster_deployment.py:52]: Why the object name is ui_driver? This can be really confusing.

File: ocs_ci/framework/conf/default_config.yaml (hunk 2 above)
Comment (shylesh) [reply ocs_ci/framework/conf/default_config.yaml:238]: @OdedViner sorry, I did it for testing but forgot to revert it

File: ocs_ci/deployment/multicluster_deployment.py (hunk 5 above)
Comment (shylesh) [reply ocs_ci/deployment/multicluster_deployment.py:52]: Because login_to_acm returns a selenium driver right ?

File: ocs_ci/ocs/constants.py [hunk 6]
```diff
@@ -11,6 +11,8 @@
 
 import os
 
+from selenium.webdriver.common.keys import Keys
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/constants.py:14]: not anymore, I will remove this.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 7]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+        # Navigate to Clusters page which has 'Create Cluster'/
+        # 'Import Cluster' buttons
+        # Here we click on "Create Cluster" and we will be in create cluster page
+        while True:
+            self.navigate_clusters_page()
+            log.info("Clicking on 'CreateCluster'")
+            if not self.check_element_presence(
+                (By.XPATH, self.acm_page_nav["cc_create_cluster"][0]), timeout=200
+            ):
+                raise ACMClusterDeployException("Create cluster button not found")
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:168]: I was combining both deployment exception with button not found, I can split it for more readability.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 8]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+            log.info("Found create cluster")
+            if not self.check_element_presence(
+                (By.XPATH, self.acm_page_nav["cc_create_cluster_index_xpath"][0]),
+                timeout=300,
+            ):
+                raise ACMClusterDeployException("Create cluster button not found")
+            log.info("Found create cluster by index path")
+            self.do_click(locator=self.acm_page_nav["cc_create_cluster"], timeout=100)
+            time.sleep(20)
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:177]: Answer to this comment and the one above: At this point during the execution I saw weird selenium behaviour, where button was found and clicked but there will be no change in the UI and for selenium its a success and run-ci will continue and fails afterwards. Basically I am using 3 different ways to ensure selenium is not fooled.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 9]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+            if self.driver.current_url.endswith("create-cluster"):
+                break
+
+    def click_next_button(self):
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:181]: I introduced this because of redability, instead of using xpath every time I am just calling this function although its single line.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 10]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+        self.do_click(self.acm_page_nav["cc_next_page_button"])
+
+    def send_keys_multiple(self, key_val):
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:184]: ```fill_multiple_textbox``` is this ok?

File: ocs_ci/ocs/ui/acm_ui.py [hunk 11]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+            locator=self.acm_page_nav["cc_infrastructure_provider_creds_dropdown"]
+        )
+        credential = format_locator(
+            self.acm_page_nav["cc_infrastructure_provider_creds_select_creds"],
+            self.platform_credential_name,
+        )
+        self.do_click(locator=credential)
+
+    @retry(ACMClusterDeployException, tries=3, delay=10, backoff=1)
+    def goto_cluster_details_page(self):
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:211]: At some places even its single line I want to convert it into function because I can then use retry decorator on that function as you know some of the problems were unsolved even using wait functions from base UI.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 12]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+    def get_deployment_status(self):
+        self.goto_cluster_details_page()
+        if self.acm_cluster_status_failed(timeout=2):
+            self.deployment_status = "failed"
+        elif self.acm_cluster_status_ready(timeout=2):
+            self.deployment_status = "ready"
+        elif self.acm_cluster_status_creating(timeout=2):
+            self.deployment_status = "creating"
+        else:
+            self.deployment_status = "unknown"
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:227]: This validation is bit tricky, lets say current status is 'Creating' then I don't want to wait for much in the 'failed' and 'ready' states. Only catch is if the status text itself takes more time to load then I could see few problems here.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 13]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+        )
+
+    def acm_cluster_status_creating(self, timeout=120):
+        return self.check_element_presence(
+            (
+                self.acm_page_nav["cc_cluster_status_page_status_creating"][1],
+                self.acm_page_nav["cc_cluster_status_page_status_creating"][0],
+            ),
+            timeout=timeout,
+        )
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:283]: As I commented earlier, I made all of these as functions so that I can easily use retries.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 14]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+        this function which iteratively fills in values from the dictionary parameter
+
+        key_val (dict): keys corresponds to the xpath of text box, value corresponds
+            to the value to be filled in
+
+        """
+        for xpath, value in key_val.items():
+            self.do_send_keys(locator=xpath, text=value)
+
+    def click_platform_and_credentials(self):
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:196]: I definitely need this function , because if I want to retry due to selenium slow response or sometimes totally misbehaving then functions makes my life easier.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 15]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+        self.navigate_create_clusters_page()
+        self.do_click(
+            locator=self.acm_page_nav[PLATFORM_XPATH_MAP[self.platform]], timeout=100
+        )
+        self.do_click(
+            locator=self.acm_page_nav["cc_infrastructure_provider_creds_dropdown"]
+        )
+        credential = format_locator(
+            self.acm_page_nav["cc_infrastructure_provider_creds_select_creds"],
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:205]: Where exactly is the extra space ? My IDE doesn't show it.

File: ocs_ci/ocs/ui/acm_ui.py [hunk 16]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+            self.platform_credential_name,
+        )
+        self.do_click(locator=credential)
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:208]: sure, I can take a call based on number of failures during run

File: ocs_ci/ocs/ui/acm_ui.py [hunk 17]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+                self.check_element_presence(
+                    (By.XPATH, self.acm_page_nav[PLATFORM_XPATH_MAP[self.platform]][0]),
+                    timeout=600,
+                )
+                break
+            except TimeoutException:
+                self.navigate_create_clusters_page()
+                self.refresh_page()
+
+        log.info("Found platform icon")
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py]: completely changed the loop

File: ocs_ci/ocs/ui/acm_ui.py [hunk 18]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+            locator=self.acm_page_nav[PLATFORM_XPATH_MAP[self.platform]], timeout=100
+        )
+
+        # "Basic vsphere credential info"
+        # 1. credential name
+        # 2. Namespace
+        # 3. Base DNS domain
+        self.do_click(locator=self.acm_page_nav["cc_provider_credentials"], timeout=100)
+        tabs = self.driver.window_handles
+        self.driver.switch_to.window(tabs[1])
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:391]: I pretty much have the same code, except that I am using indexing. I am not sure if I want to use loop as you mentioned in the example code you gave. Have a look at the latest changes and if you still feel something could brake I will change this.

File: ocs_ci/ocs/ui/acm_ui.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (am-agrawa) [reply ocs_ci/ocs/ui/acm_ui.py:183]: @OdedViner any thoughts?

File: ocs_ci/ocs/ui/acm_ui.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (am-agrawa) [reply ocs_ci/ocs/ui/acm_ui.py:187]: @OdedViner Is it okay?

File: ocs_ci/ocs/ui/acm_ui.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (am-agrawa) [reply ocs_ci/ocs/ui/acm_ui.py]: Still not very sure, maybe `send_multiple_keys` or something similar to this?

File: ocs_ci/ocs/ui/acm_ui.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (am-agrawa) [reply ocs_ci/ocs/ui/acm_ui.py:211]: I think it's not there because we have a closing bracket. My bad

File: ocs_ci/ocs/ui/acm_ui.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (am-agrawa) [reply ocs_ci/ocs/ui/acm_ui.py:214]: Ack

File: ocs_ci/ocs/ui/acm_ui.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (am-agrawa) [reply ocs_ci/ocs/ui/acm_ui.py:217]: Ack

File: ocs_ci/deployment/multicluster_deployment.py (hunk 5 above)
Comment (am-agrawa) [reply ocs_ci/deployment/multicluster_deployment.py:52]: yes, but using the `driver` word in the object name could be confusing. Just a thought

**Comment by am-agrawa** [comment]
LGTM for now, will have to observe the actual runs (whenever it happens).

**Comment by shylesh** [comment]
Rebased on latest master

**Comment by shylesh** [comment]
Regression done.

**Comment by shylesh** [comment]
At this point I was not able to test cluster 'Ready' state because of IPAM issue.

File: ocs_ci/ocs/ui/base_ui.py (hunk 3 above)
Comment (shylesh) [reply ocs_ci/ocs/ui/base_ui.py]: done

File: ocs_ci/ocs/ui/acm_ui.py [hunk 19]
```diff
@@ -103,3 +130,487 @@ def navigate_credentials_page(self):
...
+        2. Base DNS domain
+        3. Release image
+
+        """
+        release_img = self.get_ocp_release_img()
+        cluster_details = {
+            self.acm_page_nav[
+                "cc_cluster_name"
+            ]: f"{self.cluster_conf.ENV_DATA['cluster_name']}",
+            # self.acm_page_nav[
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py]: done

File: ocs_ci/ocs/ui/base_ui.py [hunk 20]
```diff
@@ -322,7 +326,7 @@ def check_element_presence(self, locator, timeout=5):
             wait = WebDriverWait(self.driver, timeout=timeout, poll_frequency=1)
             wait.until(ec.presence_of_element_located(locator))
             return True
-        except NoSuchElementException:
+        except (NoSuchElementException, TimeoutException):
             self.take_screenshot()
             logger.error("Expected element not found on UI")
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/base_ui.py:327]: done

File: ocs_ci/ocs/ui/acm_ui.py [hunk 21]
```diff
@@ -1,8 +1,35 @@
+import os
 import logging
+import time
 
+from selenium.webdriver.common.by import By
+from selenium.common.exceptions import TimeoutException
+from selenium.webdriver.common.keys import Keys
+from ocs_ci.ocs import constants
+
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py]: done

File: ocs_ci/ocs/ui/acm_ui.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py]: done

File: ocs_ci/ocs/ui/acm_ui.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py:211]: done

**Review by am-agrawa — APPROVED** [review]
Approving from my side

File: ocs_ci/ocs/ui/acm_ui.py [hunk 22]
```diff
@@ -103,3 +129,501 @@ def navigate_credentials_page(self):
...
+                self.acm_bailout_if_failed()
+        if self.acm_cluster_status_ready():
+            log.info(
+                f"Cluster create successful, Cluster {self.cluster_name} is in 'Ready' state"
+            )
+
+    def acm_bailout_if_failed(self):
+        if self.acm_cluster_status_failed():
+            raise ACMClusterDeployException("Deployment is in 'FAILED' state")
+        return
```
Comment (clacroix12) [inline ocs_ci/ocs/ui/acm_ui.py:262]: This return can be removed, it doesn't do anything

File: ocs_ci/framework/conf/default_config.yaml [hunk 23]
```diff
@@ -238,7 +238,7 @@ EXTERNAL_MODE:
 UI_SELENIUM:
   browser_type: "chrome"
   chrome_type: "google-chrome"
-  headless: True
+  headless: True
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/framework/conf/default_config.yaml]: done

File: ocs_ci/ocs/ui/base_ui.py [hunk 24]
```diff
@@ -318,14 +318,21 @@ def check_element_presence(self, locator, timeout=5):
             bool: True if the element is found, returns False otherwise and raises NoSuchElementException
 
         """
+        ret = None
```
[the remark this answers is not in the document]
Comment (shylesh) [reply ocs_ci/ocs/ui/base_ui.py]: done

File: ocs_ci/ocs/ui/acm_ui.py (hunk 22 above)
Comment (shylesh) [reply ocs_ci/ocs/ui/acm_ui.py]: done

**Comment by shylesh** [comment]
rebased on latest master

**Comment by shylesh** [comment]
Verified deployment: https://ocs4-jenkins-csb-odf-qe.apps.ocp-c1.prod.psi.redhat.com/job/qe-deploy-ocs-cluster/11704

**Review by petr-balogh — APPROVED** [review]

**Review by prsurve — APPROVED** [review]
